### ISSUE TITLE
should oboe::DataCallbackResult::Stop also stop the stream?

### ISSUE THREAD

**Issue by SchmadenSchmuki** [google/oboe#277]
Hello
I would like to know, what to expect when I call oboe::DataCallbackResult::Stop in the onAudioReady-method?

In my recordingcode I use it as returnvalue when I have no memory left to store additional data.
The result is that onAudioReady still gets called continuously (and returns oboe::DataCallbackResult::Stop) until the stream is stopped somewhere else.

I can replace oboe::DataCallbackResult::Stop with oboeStream->stop() but I was wondering whether oboe::DataCallbackResult::Stop does something important and should also be called?

**Comment by philburk** [comment]
Thanks for reporting this. It should stop the callbacks from being called. 

I cannot reproduce this bug.
What OS version were you using?
What device?

Can you insert a few lines from the logcat around the time the Stop is returned?

Are you seeing any logs like this:
    AAudioServiceStreamBase: writeUpMessageQueue(): Queue full. Did client die? Shared

> I can replace oboe::DataCallbackResult::Stop with oboeStream->stop()

No. Do not call oboeStream->stop() from the callback. The stop() has to be called from another thread. If I can reproduce the bug then I may change Oboe so that it stops the stream from another thread if this problem occurs.

**Comment by SchmadenSchmuki** [comment]
I use a Nexus 5X which runs Android 8.1

logs from starting the stream are as following:

    2018-11-21 19:05:59.434 2862-9874/? I/EventLogSendingHelper: Sending log events.
    2018-11-21 19:06:01.846 10602-10602/jogi.programme.android.soundtests D/AAudio: AAudioStream_requestStart(0x7147eac500) called --------------
    2018-11-21 19:06:01.851 10602-10602/jogi.programme.android.soundtests D/AAudio: AAudioStream_requestStart(0x7147eac500) returned 0 ---------
    2018-11-21 19:06:01.851 10602-10614/jogi.programme.android.soundtests D/AudioStreamLegacy: onAudioDeviceUpdate() deviceId 8
    2018-11-21 19:06:01.852 562-10657/? I/msm8974_platform: platform_check_capture_backend_cfg:txbecf: afe: Codec backend needs to be updated. new bit width: 0 new sample rate: 0 new channel: 2
    2018-11-21 19:06:01.853 562-10657/? D/sound_trigger_platform: platform_stdev_check_and_update_concurrency: concurrency active 0, tx 1, rx 0, concurrency session_allowed 0
    2018-11-21 19:06:01.853 562-10657/? D/audio_hw_primary: enable_snd_device: snd_device(75: voice-rec-mic)
    2018-11-21 19:06:01.853 562-10657/? D/audio_route: Apply path: voice-rec-mic
    2018-11-21 19:06:01.858 562-10657/? D/audio_hw_primary: enable_audio_route: usecase(10) apply and update mixer path: low-latency-record
    2018-11-21 19:06:01.859 562-10657/? D/audio_route: Apply path: low-latency-record
    2018-11-21 19:06:03.897 10602-10639/jogi.programme.android.soundtests I/Callback: log before oboe::DataCallbackResult::Stop is called, times already called: 0
    2018-11-21 19:06:03.908 10602-10639/jogi.programme.android.soundtests I/Callback: log before oboe::DataCallbackResult::Stop is called, times already called: 1
    2018-11-21 19:06:03.918 10602-10639/jogi.programme.android.soundtests I/Callback: log before oboe::DataCallbackResult::Stop is called, times already called: 2
    2018-11-21 19:06:03.928 10602-10639/jogi.programme.android.soundtests I/Callback: log before oboe::DataCallbackResult::Stop is called, times already called: 3
    2018-11-21 19:06:03.939 10602-10639/jogi.programme.android.soundtests I/Callback: log before oboe::DataCallbackResult::Stop is called, times already called: 4
    2018-11-21 19:06:03.949 10602-10639/jogi.programme.android.soundtests I/Callback: log before oboe::DataCallbackResult::Stop is called, times already called: 5
    2018-11-21 19:06:03.960 10602-10639/jogi.programme.android.soundtests I/Callback: log before oboe::DataCallbackResult::Stop is called, times already called: 6
    2018-11-21 19:06:03.970 10602-10639/jogi.programme.android.soundtests I/Callback: log before oboe::DataCallbackResult::Stop is called, times already called: 7
    2018-11-21 19:06:03.985 10602-10639/jogi.programme.android.soundtests I/Callback: log before oboe::DataCallbackResult::Stop is called, times already called: 8
    2018-11-21 19:06:03.996 10602-10639/jogi.programme.android.soundtests I/Callback: log before oboe::DataCallbackResult::Stop is called, times already called: 9
    2018-11-21 19:06:04.006 10602-10639/jogi.programme.android.soundtests I/Callback: log before oboe::DataCallbackResult::Stop is called, times already called: 10

sometime later

     2018-11-21 19:06:04.584 562-10642/? W/AudioFlinger: overrun on read from pipe

appears between those lines, but I guess that's because oboe::DataCallbackResult::Continue isn't returned anymore.
The time until the first stop call is produced is 2 seconds because that's the set maximum recordingtime.

I can't see any other error logs of use or like the one you've mentioned, there are also no problems with the data I receive. Nonetheless, the stream doesn't stop.
Calling stop() from outside works fine.

**Comment by philburk** [comment]
THANKS for the extra info. I am now able to reproduce the bug. 
It happens on AAudio Legacy streams but not on MMAP streams. 
It happens on 8.1 (OC_MR1) but not on 9.0 (Pie).
It does not happen on OpenSL ES.

At a minimum, Oboe should block the callback from being called if the callback has requested Stop.
On 9.0, the Stop causes some log spew so I may spawn a thread to requestStop() on the stream.

**Comment by philburk** [comment]
Fixed in Pie in https://googleplex-android-review.git.corp.google.com/c/platform/frameworks/av/+/3340045

**Comment by SchmadenSchmuki** [comment]
Thanks for the info.
Changed the stream to oboe::AudioApi::OpenSLES.
Resulting logs are as following:

    2018-11-22 01:48:18.054 14924-14924/jogi.programme.android.soundtests D/OboeAudio: AudioInputStreamOpenSLES::requestStart()
    2018-11-22 01:48:18.054 14924-14924/jogi.programme.android.soundtests D/OboeAudio: AudioInputStreamOpenSLES::setRecordState(3)
    2018-11-22 01:48:18.065 562-14973/? I/msm8974_platform: platform_check_capture_backend_cfg:txbecf: afe: Codec backend needs to be updated. new bit width: 0 new sample rate: 0 new channel: 2
    2018-11-22 01:48:18.065 562-14973/? D/sound_trigger_platform: platform_stdev_check_and_update_concurrency: concurrency active 0, tx 1, rx 0, concurrency session_allowed 0
    2018-11-22 01:48:18.065 562-14973/? D/audio_hw_primary: enable_snd_device: snd_device(75: voice-rec-mic)
    2018-11-22 01:48:18.065 562-14973/? D/audio_route: Apply path: voice-rec-mic
    2018-11-22 01:48:18.070 562-14973/? D/audio_hw_primary: enable_audio_route: usecase(10) apply and update mixer path: low-latency-record
    2018-11-22 01:48:18.071 562-14973/? D/audio_route: Apply path: low-latency-record
    2018-11-22 01:48:18.113 14924-14956/jogi.programme.android.soundtests D/OboeAudio: AudioStream::fireCallback() scheduler = SCHED_FIFO
    2018-11-22 01:48:20.114 14924-14956/jogi.programme.android.soundtests I/Callback: log before oboe::DataCallbackResult::Stop is called, times already called: 0
    2018-11-22 01:48:20.114 14924-14956/jogi.programme.android.soundtests E/OboeAudio: Oboe callback returned 1
    2018-11-22 01:48:21.569 562-14959/? W/AudioFlinger: overrun on read from pipe
    2018-11-22 01:48:24.341 562-14959/? W/AudioFlinger: overrun on read from pipe
    2018-11-22 01:48:24.766 14924-14924/jogi.programme.android.soundtests I/info: pStreamRecording->getState() as number: 4
    2018-11-22 01:48:24.766 14924-14924/jogi.programme.android.soundtests I/info: oboeStreamState::Started as number: 4
    2018-11-22 01:48:24.766 14924-14924/jogi.programme.android.soundtests I/info: oboeStreamState::Stopped as number: 10
    2018-11-22 01:48:26.395 562-14959/? W/AudioFlinger: overrun on read from pipe
    2018-11-22 01:48:30.023 562-14959/? I/chatty: uid=1041(audioserver) AudioIn_19E identical 2 lines
    2018-11-22 01:48:31.813 562-14959/? W/AudioFlinger: overrun on read from pipe
    2018-11-22 01:48:31.849 14924-14924/jogi.programme.android.soundtests D/OboeAudio: AudioInputStreamOpenSLES::requestStop()
    2018-11-22 01:48:31.849 14924-14924/jogi.programme.android.soundtests D/OboeAudio: AudioInputStreamOpenSLES::setRecordState(1)
    2018-11-22 01:48:31.895 562-14959/? D/audio_hw_primary: disable_audio_route: usecase(10) reset and update mixer path: low-latency-record

So the callback wasn't called anymore but the stream remained in the started state (logs at 2018-11-22 01:48:24.766) and "W/AudioFlinger: overrun on read from pipe" kept coming until I stopped the stream manually.

**Comment by weizhang1999** [comment]
Hello,
I have a similar problem. After "return DataCallbackResult::Stop" in onAudioReady, the stream->getState() always returns 4 (StreamState::Started). Are we supposed to get StreamState::Stopped instead?

Can you please clarify when I should "return DataCallbackResult::Stop" in onAudioReady?
For example,
If I have the last buffer to play, can I copy the data to void* audioData and return DataCallbackResult::Stop?
Or do I need to wait until there is nothing left before returning DataCallbackResult::Stop?
In other words,
Will the data copied to audioData be processed when returning DataCallbackResult::Stop?

Thanks!

**Comment by dturner** [comment]
@weizhang1999 Please post your question/issue in a new issue and make sure you include the phone model and android version you are using.

**Comment by dturner** [comment]
I'm running into this issue on both Pixel 2 (PQ1A.181105.017.A1) and Pixel XL (PQ1A.181105.014) running Android 9.0. 

I've created a [modified version of the MegaDrone sample here](https://github.com/google/oboe/tree/instrumentedcallback) which waits until 2500 callbacks have occurred (4 seconds at 96 frame burst and 48000 frames/sec) before returning `DataCallbackResult::Stop`:



Steps to repro: 

- Run the MegaDrone sample
- Tap and hold the screen for 4 seconds
- You should hear a sound being played

Expected result 
- The sound stops after ~4 seconds

Actual result
- The audio starts to glitch after ~seconds, the same buffer of audio is played over and over, like a skipping CD
- After a while the app crashes

Pixel XL logs show: 

    2018-12-04 15:25:19.147 665-22400/? E/AAudioServiceStreamBase: writeUpMessageQueue(): Queue full. Did client die? MMAP

Pixel 2 logs here: 

2018-12-04 15:21:19.204 32174-32174/com.example.oboe.megadrone D/AAudio: AAudioStreamBuilder_openStream() called ----------------------------------------
2018-12-04 15:21:19.204 32174-32174/com.example.oboe.megadrone I/AAudioStream: open() rate   = 0, channels    = 0, format   = 0, sharing = EX, dir = OUTPUT
2018-12-04 15:21:19.204 32174-32174/com.example.oboe.megadrone I/AAudioStream: open() device = 0, sessionId   = 0, perfMode = 12, callback: ON with frames = 0
2018-12-04 15:21:19.204 32174-32174/com.example.oboe.megadrone I/AAudioStream: open() usage  = 1, contentType = 2, inputPreset = 6
2018-12-04 15:21:19.210 32174-32174/com.example.oboe.megadrone D/AAudio: AAudioStreamBuilder_openStream() returns 0 = AAUDIO_OK for (0x794344e800) ----------------
2018-12-04 15:21:19.210 32174-32174/com.example.oboe.megadrone D/OboeAudio: AudioStreamAAudio.open() app    format = 2
2018-12-04 15:21:19.210 32174-32174/com.example.oboe.megadrone D/OboeAudio: AudioStreamAAudio.open() sample rate   = 48000
2018-12-04 15:21:19.210 32174-32174/com.example.oboe.megadrone D/OboeAudio: AudioStreamAAudio.open() capacity      = 3072
2018-12-04 15:21:19.210 32174-32174/com.example.oboe.megadrone D/OboeAudio: AudioStreamAAudio.open: AAudioStream_Open() returned AAUDIO_OK, mAAudioStream = 0x794344e800
2018-12-04 15:21:19.211 32174-32174/com.example.oboe.megadrone D/AudioStreamInternal_Client: setBufferSize() req = 192 => 192
2018-12-04 15:21:19.211 32174-32174/com.example.oboe.megadrone D/AAudio: AAudioStream_requestStart(0x794344e800) called --------------
2018-12-04 15:21:19.220 32174-32174/com.example.oboe.megadrone D/AAudio: AAudioStream_requestStart(0x794344e800) returned 0 ---------
2018-12-04 15:21:19.220 32174-32198/com.example.oboe.megadrone D/AudioStreamInternalPlay_Client: callbackLoop() entering >>>>>>>>>>>>>>>
2018-12-04 15:21:19.220 32174-32198/com.example.oboe.megadrone V/OboeAudio: CPU ID 4 added to cores set
2018-12-04 15:21:19.220 32174-32198/com.example.oboe.megadrone V/OboeAudio: CPU ID 5 added to cores set
2018-12-04 15:21:19.220 32174-32198/com.example.oboe.megadrone V/OboeAudio: Thread affinity set
2018-12-04 15:21:19.222 32174-32198/com.example.oboe.megadrone D/AudioStreamInternal_Client: onEventFromServer - AAUDIO_SERVICE_EVENT_VOLUME 0.096924
2018-12-04 15:21:19.222 32174-32198/com.example.oboe.megadrone D/AudioStreamInternalPlay_Client: doSetVolume() mStreamVolume * duckAndMuteVolume = 0.096924 * 1.000000 = 0.096924
2018-12-04 15:21:19.222 32174-32198/com.example.oboe.megadrone D/AudioStreamInternal_Client: onEventFromServer - AAUDIO_SERVICE_EVENT_VOLUME 0.096924
2018-12-04 15:21:19.222 32174-32198/com.example.oboe.megadrone D/AudioStreamInternalPlay_Client: doSetVolume() mStreamVolume * duckAndMuteVolume = 0.096924 * 1.000000 = 0.096924
2018-12-04 15:21:19.222 32174-32198/com.example.oboe.megadrone D/AudioStreamInternal_Client: onEventFromServer - got AAUDIO_SERVICE_EVENT_STARTED
2018-12-04 15:21:19.222 32174-32198/com.example.oboe.megadrone D/AudioStreamInternal_Client: onEventFromServer - AAUDIO_SERVICE_EVENT_VOLUME 0.096924
2018-12-04 15:21:19.222 32174-32198/com.example.oboe.megadrone D/AudioStreamInternalPlay_Client: doSetVolume() mStreamVolume * duckAndMuteVolume = 0.096924 * 1.000000 = 0.096924
2018-12-04 15:21:19.228 32174-32174/com.example.oboe.megadrone D/OpenGLRenderer: Skia GL Pipeline
2018-12-04 15:21:19.267 32174-32199/com.example.oboe.megadrone I/Adreno: QUALCOMM build                   : 7bf2852, Ie32bfa6f6f
    Build Date                       : 06/03/18
    OpenGL ES Shader Compiler Version: EV031.24.00.00
    Local Branch                     : googdrp
    Remote Branch                    : 
    Remote Branch                    : 
    Reconstruct Branch               : 
2018-12-04 15:21:19.267 32174-32199/com.example.oboe.megadrone I/Adreno: Build Config                     : S L 4.0.10 AArch64
2018-12-04 15:21:19.269 32174-32199/com.example.oboe.megadrone I/Adreno: PFP: 0x005ff110, ME: 0x005ff066
2018-12-04 15:21:19.273 32174-32199/com.example.oboe.megadrone I/ConfigStore: android::hardware::configstore::V1_0::ISurfaceFlingerConfigs::hasWideColorDisplay retrieved: 1
2018-12-04 15:21:19.273 32174-32199/com.example.oboe.megadrone I/ConfigStore: android::hardware::configstore::V1_0::ISurfaceFlingerConfigs::hasHDRDisplay retrieved: 0
2018-12-04 15:21:19.273 32174-32199/com.example.oboe.megadrone I/OpenGLRenderer: Initialized EGL, version 1.4
2018-12-04 15:21:19.273 32174-32199/com.example.oboe.megadrone D/OpenGLRenderer: Swap behavior 1
2018-12-04 15:21:29.219 32174-32198/com.example.oboe.megadrone D/OboeAudio: Returning STOP from callback
2018-12-04 15:21:29.219 32174-32198/com.example.oboe.megadrone D/AudioStreamInternalPlay_Client: callbackLoop() exiting, result = 96, isActive() = 1 <<<<<<<<<<<<<<
2018-12-04 15:22:00.729 32174-32174/com.example.oboe.megadrone D/AAudio: AAudioStream_close(0x794344e800) called ---------------
2018-12-04 15:22:00.729 32174-32174/com.example.oboe.megadrone D/AudioStreamInternal_Client: close(): mServiceStreamHandle = 0x00000024
2018-12-04 15:22:00.864 32174-32174/com.example.oboe.megadrone D/AAudioStream: destroying 0x794344e800, state = AAUDIO_STREAM_STATE_CLOSED
2018-12-04 15:22:00.864 32174-32174/com.example.oboe.megadrone D/AAudio: AAudioStream_close(0x794344e800) returned 0 ---------
2018-12-04 15:22:13.298 32174-32174/com.example.oboe.megadrone D/OboeAudio: AudioStreamAAudio() call isSupported()
2018-12-04 15:22:13.299 32174-32174/com.example.oboe.megadrone D/AAudio: AAudioStreamBuilder_openStream() called ----------------------------------------
2018-12-04 15:22:13.299 32174-32174/com.example.oboe.megadrone I/AAudioStream: open() rate   = 0, channels    = 0, format   = 0, sharing = EX, dir = OUTPUT
2018-12-04 15:22:13.299 32174-32174/com.example.oboe.megadrone I/AAudioStream: open() device = 0, sessionId   = 0, perfMode = 12, callback: ON with frames = 0
2018-12-04 15:22:13.299 32174-32174/com.example.oboe.megadrone I/AAudioStream: open() usage  = 1, contentType = 2, inputPreset = 6
2018-12-04 15:22:13.335 32174-32174/com.example.oboe.megadrone D/AAudio: AAudioStreamBuilder_openStream() returns 0 = AAUDIO_OK for (0x794344e800) ----------------
2018-12-04 15:22:13.338 32174-32174/com.example.oboe.megadrone D/OboeAudio: AudioStreamAAudio.open() app    format = 2
2018-12-04 15:22:13.338 32174-32174/com.example.oboe.megadrone D/OboeAudio: AudioStreamAAudio.open() sample rate   = 48000
2018-12-04 15:22:13.338 32174-32174/com.example.oboe.megadrone D/OboeAudio: AudioStreamAAudio.open() capacity      = 3072
2018-12-04 15:22:13.338 32174-32174/com.example.oboe.megadrone D/OboeAudio: AudioStreamAAudio.open: AAudioStream_Open() returned AAUDIO_OK, mAAudioStream = 0x794344e800
2018-12-04 15:22:13.339 32174-32174/com.example.oboe.megadrone A/libc: Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x8a9f804c70 in tid 32174 (.oboe.megadrone), pid 32174 (.oboe.megadrone)

**Comment by philburk** [comment]
@Don - I think you may have found a separate but related bug. The original bug was on the 5X so it was on the Legacy path. Your crash is on a Pixel 2 on the MMAP data path.

I think a way to summarize the steps for Don's crash are:

- callback returns Stop code on MMAP stream
- stop() is **not** called
- app calls close
- app calls open
- crash

**Comment by philburk** [comment]
There is a possible fix in https://github.com/google/oboe/pull/323
When a stop is requested from a callback, it launches a thread to do the stop.
It also adds some locks to prevent a collision if the app also does a stop.
It needs a through review.
Also please try the patch to see if it helps with this issue.

**Comment by SchmadenSchmuki** [comment]
Hello
Downloaded the latest oboe-version today and got the following results (Nexus 5X, Android 8.1):

OpenSLES: 
the callback wasn't called anymore but the stream remained in the started state until I stopped it manually.
AAudio:
the callback wasn't stopped, so it was called repeatedly.

The results were the same for input- and outputstream.

I just did a simple test by placing a log and oboe::DataCallbackResult::Stop as first instructions in the respective callbacks. The code that worked fine otherwise, before the changes (as far as I could see) .

**Comment by philburk** [comment]
> Downloaded the latest oboe-version today and got the following results

I think the fix was merged the day after your test. Can you please update Oboe again and verify that you have the commits in #323.  If it works I will close this issue.

**Comment by SchmadenSchmuki** [comment]
Thanks for the info, downloaded the latest version an repeated the simple test.

OpenSLES:
Everything worked like a charm, after oboe::DataCallbackResult::Stop the callback wasn't called anymore and the streamstate changed to stopped. (input and output)

AAudio:
Output: everything was fine.
Input:
After oboe::DataCallbackResult::Stop the callback wasn't called anymore in all cases.
The streamstate changed to stopped only the first time. 
When calling start on the same stream again and returning oboe::DataCallbackResult::Stop again from the callback, the stream changed to and remained in the state stopping.
I don't know if that would cause a problem but it might lead to some changes and that's why I haven't tested the usual occurances of oboe::DataCallbackResult::Stop in my code.

**Comment by SchmadenSchmuki** [comment]
Downloaded the latest oboe and tested it in my normal app-routine (Nexus 5X, Android 8.1).

Again, everything fine when using OpenSL ES,
when using AAudio, the streams (in and out) sometimes remained in state stopping, sometimes stopped.
Couldn't really find a reason for that, so it could also be caused by my app.
The only thing I could observe was:
When the log
       
    D/OboeAudio: oboe_stop_thread_proc() returning (((((----

wasn't written, the streams always remained in stopping.
(The log appeared 2-3 times for the outputstream, when 3 times the stream got to state stopped and when 2 times it was sometimes stopping, sometimes stopped. For the inputstream the log appeared 2 times with stopped and not at all with stopping.)

If noone else can verify that behaviour, then it is probably only in my app.
Also, when the stream remains in stopping, it doesn't cause me any trouble (and for now I won't be looking for the error), so as far as I'm concerned the issue can be closed.

**Comment by philburk** [comment]
Thanks for reporting this!
